File: python/백준/구현/20055 컨베이어 벨트 위의 로봇.py
N, K = map(int, input().split())

belt = list(map(int, input().split()))

robots = [False] * N  

step = 0

while True:
    step += 1 
    
    belt = [belt[-1]] + belt[:-1]
    robots = [False] + robots[:-1]
    robots[-1] = False
    
    for i in range(N - 2, -1, -1):
        if robots[i] and not robots[i + 1] and belt[i + 1] > 0:
            robots[i] = False
            robots[i + 1] = True
            belt[i + 1] -= 1
    
    robots[-1] = False
    
    if belt[0] > 0:
        robots[0] = True
        belt[0] -= 1
    
    if belt.count(0) >= K:
        break
    
print(step)

File: cpp/백준/구현/20055 컨베이어 벨트 위의 로봇.cpp
#include <iostream>
#include <vector>
#include <deque>
#include <algorithm>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int N, K;
    cin >> N >> K;

    deque<int> belt(2 * N);
    for (int i = 0; i < 2 * N; ++i) cin >> belt[i];

    deque<bool> robots(N, false);

    int step = 0;

    while (true) {
        step++;

        int last = belt.back();
        belt.pop_back();
        belt.push_front(last);

        robots.pop_back();
        robots.push_front(false);
        robots[N - 1] = false;

        for (int i = N - 2; i >= 0; --i) {
            if (robots[i] && !robots[i + 1] && belt[i + 1] > 0) {
                robots[i] = false;
                robots[i + 1] = true;
                belt[i + 1] -= 1;
            }
        }

        robots[N - 1] = false;

        if (belt[0] > 0) {
            robots[0] = true;
            belt[0] -= 1;
        }

        int zero_cnt = 0;
        for (int v : belt) if (v == 0) zero_cnt++;
        if (zero_cnt >= K) break;
    }

    cout << step << "\n";
    return 0;
}
